File: Number_of_Increasing_Paths_in_a_Grid.py
class Solution:
    def countPaths(self, grid):
        rows, cols = len(grid), len(grid[0])
        mod = 10 ** 9 + 7
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        dp = [[0] * cols for _ in range(rows)]

        def dfs(r, c):
            if dp[r][c]:
                return dp[r][c]
            
            paths = 1

            for dr, dc in directions:
                prev_r, prev_c = r + dr, c + dc
                if (0 <= prev_r < rows and
                0 <= prev_c < cols and
                grid[prev_r][prev_c] < grid[r][c]):

                    paths += dfs(prev_r, prev_c) % mod
            
            dp[r][c] = paths

            return paths

        path_count = 0
        for r in range(rows):
            for c in range(cols):
                path_count += dfs(r, c)
        
        return path_count % mod

File: Number_of_Increasing_Paths_in_a_Grid.java
class Solution {
    public int countPaths(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int mod = (int) Math.pow(10, 9) + 7;
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        int[][] dp = new int[rows][cols];
        for (int[] row : dp) {
            Arrays.fill(row, 0);
        }

        int pathCount = 0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                pathCount = (pathCount + dfs(r, c, grid, dp, directions, mod)) % mod;
            }
        }

        return pathCount;        
    }

    private int dfs(int r, int c, int[][] grid, int[][] dp, int[][] directions, int mod) {
        int rows = grid.length;
        int cols = grid[0].length;

        if (dp[r][c] > 0) {
            return dp[r][c];
        }

        int paths = 1;
        for (int[] dir : directions) {
            int dr = dir[0];
            int dc = dir[1];
            int prevR = r + dr;
            int prevC = c + dc;

            if (prevR >= 0 && prevR < rows && prevC >= 0 && prevC < cols && grid[prevR][prevC] < grid[r][c]) {
                paths = (paths + dfs(prevR, prevC, grid, dp, directions, mod)) % mod;
            }
        }

        dp[r][c] = paths;
        return paths;
    }    
}
